agent/wsclient/client.go

```diff
@@ -0,0 +1,268 @@
+	wsConn, err := tls.DialWithDialer(timeoutDialer, "tcp", dialHost, &tls.Config{InsecureSkipVerify: cs.AcceptInvalidCert})
+	if err != nil {
+		return err
+	}
+
+	websocketConn, httpResponse, err := websocket.NewClient(wsConn, parsedURL, request.Header, readBufSize, writeBufSize)
+	if httpResponse != nil {
+		defer httpResponse.Body.Close()
+	}
+	if err != nil {
+		var resp []byte
+		if httpResponse != nil {
+			var readErr error
+			resp, readErr = ioutil.ReadAll(httpResponse.Body)
+			if readErr != nil {
+				return errors.New("Unable to read websocket connection: " + readErr.Error() + ", " + err.Error())
+			}
+			// If there's a response, we can try to unmarshal it into one of the
+			// modeled error types
+			possibleError, _, decodeErr := DecodeData(resp, cs.TypeDecoder)
+			if decodeErr == nil {
+				return cs.NewError(possibleError)
+			}
+		}
+		log.Warn("Error creating a websocket client", "err", err)
+		return errors.New(string(resp) + ", " + err.Error())
+	}
+	cs.Conn = websocketConn
+	return nil
+}
+
+// AddRequestHandler adds a request handler to this client.
+// A request handler *must* be a function taking a single argument, and that
+// argument *must* be a pointer to a recognized 'ecsacs' struct.
+// E.g. if you desired to handle messages from acs of type 'FooMessage', you
+// would pass the following handler in:
+//     func(message *ecsacs.FooMessage)
+// This function will panic if the passed in function does not have one pointer
+// argument or the argument is not a recognized type.
+// Additionally, the request handler will block processing of further messages
+// on this connection so it's important that it return quickly.
+func (cs *ClientServerImpl) AddRequestHandler(f RequestHandler) {
+	firstArg := reflect.TypeOf(f).In(0)
+	firstArgTypeStr := firstArg.Elem().Name()
+	recognizedTypes := cs.GetRecognizedTypes()
+	_, ok := recognizedTypes[firstArgTypeStr]
+	if !ok {
+		panic("AddRequestHandler called with invalid function; argument type not recognized: " + firstArgTypeStr)
+	}
+	cs.RequestHandlers[firstArgTypeStr] = f
+}
+
+// MakeRequest makes a request using the given input. Note, the input *MUST* be
+// a pointer to a valid backend type that this client recognises
+func (cs *ClientServerImpl) MakeRequest(input interface{}) error {
+	send, err := cs.CreateRequestMessage(input)
+	if err != nil {
+		return err
+	}
+
+	// Over the wire we send something like
+	// {"type":"AckRequest","message":{"messageId":"xyz"}}
+	return cs.Conn.WriteMessage(websocket.TextMessage, send)
+}
+
+// ConsumeMessages reads messages from the websocket connection and handles read
+// messages from an active connection.
+func (cs *ClientServerImpl) ConsumeMessages() error {
+	var err error
+	for {
+		messageType, message, cerr := cs.Conn.ReadMessage()
+		err = cerr
+		if err != nil {
+			if message != nil {
+				log.Error("Error getting message from acs", "err", err, "message", message)
+			} else {
```

Comment:
this is missing the 'io.EOF' behavior that's there in the previous 'Serve' code. This will result in lots of log.Error level output (one per ACS disconnect). Mind double checking that there's not any other differences between these two chunks as well?